File: android/rctmgl/src/main/java-mapboxgl/common/com/mapbox/rctmgl/components/camera/CameraStop.java
package com.mapbox.rctmgl.components.camera;

import android.content.Context;
import androidx.annotation.NonNull;
import android.util.DisplayMetrics;

import com.facebook.react.bridge.ReadableMap;
import com.mapbox.geojson.FeatureCollection;
import com.mapbox.geojson.Point;
import com.mapbox.mapboxsdk.camera.CameraPosition;
import com.mapbox.mapboxsdk.camera.CameraUpdate;
import com.mapbox.mapboxsdk.camera.CameraUpdateFactory;
import com.mapbox.mapboxsdk.geometry.LatLng;
import com.mapbox.mapboxsdk.geometry.LatLngBounds;
import com.mapbox.mapboxsdk.maps.MapboxMap;
import com.mapbox.rctmgl.components.camera.constants.CameraMode;
import com.mapbox.rctmgl.utils.GeoJSONUtils;

import com.mapbox.rctmgl.components.mapview.RCTMGLMapView;

/**
 * Created by nickitaliano on 9/5/17.
 */

public class CameraStop {
    private Double mBearing;
    private Double mTilt;
    private Double mZoom;
    private LatLng mLatLng;

    private LatLngBounds mBounds;
    private int mPaddingLeft = 0;
    private int mPaddingRight = 0;
    private int mPaddingBottom = 0;
    private int mPaddingTop = 0;

    private int mMode = CameraMode.EASE;
    private int mDuration = 2000;
    private MapboxMap.CancelableCallback mCallback;

    public CameraStop() {
    }

    public void setBearing(double bearing) {
        mBearing = bearing;
    }

    public void setTilt(double tilt) {
        mTilt = tilt;
    }

    public void setZoom(double zoom) {
        mZoom = zoom;
    }

    public void setLatLng(LatLng latLng) {
        mLatLng = latLng;
    }

    public void setDuration(int duration) {
        mDuration = duration;
    }

    public void setCallback(MapboxMap.CancelableCallback callback) {
        mCallback = callback;
    }

    public void setBounds(LatLngBounds bounds) {
        mBounds = bounds;
    }

    public void setPadding(int paddingLeft, int paddingRight, int paddingTop, int paddingBottom) {
        mPaddingLeft = paddingLeft;
        mPaddingRight = paddingRight;
        mPaddingTop = paddingTop;
        mPaddingBottom = paddingBottom;
    }

    public void setMode(@CameraMode.Mode int mode) {
        mMode = mode;
    }

    public CameraUpdateItem toCameraUpdate(RCTMGLMapView mapView) {
        MapboxMap map = mapView.getMapboxMap();
        CameraPosition currentCamera = map.getCameraPosition();
        CameraPosition.Builder builder = new CameraPosition.Builder(currentCamera);

        // Adding map padding to the camera padding which is the same behavior as
        // mapbox native does on iOS
        double[] contentInset = mapView.getContentInset();

        int paddingLeft = Double.valueOf(contentInset[0] + mPaddingLeft).intValue();
        int paddingTop = Double.valueOf(contentInset[1] + mPaddingTop).intValue();
        int paddingRight = Double.valueOf(contentInset[2] + mPaddingRight).intValue();
        int paddingBottom = Double.valueOf(contentInset[3] + mPaddingBottom).intValue();

        int[] cameraPadding = {paddingLeft, paddingTop, paddingRight, paddingBottom};
        int[] cameraPaddingClipped = clippedPadding(cameraPadding, mapView);

        boolean hasSetZoom = false;

        if (mLatLng != null) {
            builder.target(mLatLng);
            builder.padding(
                cameraPaddingClipped[0],
                cameraPaddingClipped[1],
                cameraPaddingClipped[2],
                cameraPaddingClipped[3]
            );
        } else if (mBounds != null) {
            double tilt = mTilt != null ? mTilt : currentCamera.tilt;
            double bearing = mBearing != null ? mBearing : currentCamera.bearing;

            CameraPosition boundsCamera = map.getCameraForLatLngBounds(mBounds, cameraPaddingClipped, bearing, tilt);
            if (boundsCamera != null) {
                builder.target(boundsCamera.target);
                builder.zoom(boundsCamera.zoom);
                builder.padding(boundsCamera.padding);
            } else {
                CameraUpdate update = CameraUpdateFactory.newLatLngBounds(
                    mBounds,
                    cameraPaddingClipped[0],
                    cameraPaddingClipped[1],
                    cameraPaddingClipped[2],
                    cameraPaddingClipped[3]
                );
                return new CameraUpdateItem(map, update, mDuration, mCallback, mMode);
            }
            hasSetZoom = true;
        }

        if (mBearing != null) {
            builder.bearing(mBearing);
        }

        if (mTilt != null) {
            builder.tilt(mTilt);
        }

        if (mZoom != null && !hasSetZoom) {
            builder.zoom(mZoom);
        }

        return new CameraUpdateItem(map, CameraUpdateFactory.newCameraPosition(builder.build()), mDuration, mCallback, mMode);
    }

    public static CameraStop fromReadableMap(Context context, @NonNull ReadableMap readableMap, MapboxMap.CancelableCallback callback) {
        CameraStop stop = new CameraStop();

        if (readableMap.hasKey("pitch")) {
            stop.setTilt(readableMap.getDouble("pitch"));
        }

        if (readableMap.hasKey("heading")) {
            stop.setBearing(readableMap.getDouble("heading"));
        }

        int paddingTop = getPaddingByKey(readableMap, "paddingTop");
        int paddingRight = getPaddingByKey(readableMap, "paddingRight");
        int paddingBottom = getPaddingByKey(readableMap, "paddingBottom");
        int paddingLeft = getPaddingByKey(readableMap, "paddingLeft");

        // scale padding by pixel ratio
        DisplayMetrics metrics = context.getResources().getDisplayMetrics();
        paddingTop = Float.valueOf(paddingTop * metrics.scaledDensity).intValue();
        paddingRight = Float.valueOf(paddingRight * metrics.scaledDensity).intValue();
        paddingBottom = Float.valueOf(paddingBottom * metrics.scaledDensity).intValue();
        paddingLeft = Float.valueOf(paddingLeft * metrics.scaledDensity).intValue();

        stop.setPadding(
                paddingLeft,
                paddingRight,
                paddingTop,
                paddingBottom
        );

        if (readableMap.hasKey("centerCoordinate")) {
            Point target = GeoJSONUtils.toPointGeometry(readableMap.getString("centerCoordinate"));
            stop.setLatLng(GeoJSONUtils.toLatLng(target));
        }

        if (readableMap.hasKey("zoom")) {
            stop.setZoom(readableMap.getDouble("zoom"));
        }

        if (readableMap.hasKey("duration")) {
            stop.setDuration(readableMap.getInt("duration"));
        }

        if (readableMap.hasKey("bounds")) {
            FeatureCollection collection = FeatureCollection.fromJson(readableMap.getString("bounds"));
            stop.setBounds(GeoJSONUtils.toLatLngBounds(collection));
        }

        if (readableMap.hasKey("mode")) {
            switch (readableMap.getInt("mode")) {
                case CameraMode.FLIGHT:
                    stop.setMode(CameraMode.FLIGHT);
                    break;
                case CameraMode.LINEAR:
                    stop.setMode(CameraMode.LINEAR);
                    break;
                case CameraMode.NONE:
                    stop.setMode(CameraMode.NONE);
                    break;
                default:
                    stop.setMode(CameraMode.EASE);
            }
        }

        stop.setCallback(callback);
        return stop;
    }

    private static int[] clippedPadding(int[] padding, RCTMGLMapView mapView) {
        int mapHeight = mapView.getHeight();
        int mapWidth = mapView.getWidth();

        int left = padding[0];
        int top = padding[1];
        int right = padding[2];
        int bottom = padding[3];

        int resultLeft = left;
        int resultTop = top;
        int resultRight = right;
        int resultBottom = bottom;

        if (top + bottom >= mapHeight) {
            double totalPadding = top + bottom;
            double extra = totalPadding - mapHeight + 1.0; // add 1 to compensate for floating point math
            resultTop -= (top * extra) / totalPadding;
            resultBottom -= (bottom * extra) / totalPadding;
        }

        if (left + right >= mapWidth) {
            double totalPadding = left + right;
            double extra = totalPadding - mapWidth + 1.0; // add 1 to compensate for floating point math
            resultLeft -= (left * extra) / totalPadding;
            resultRight -= (right * extra) / totalPadding;
        }

        return new int[] {resultLeft, resultTop, resultRight, resultBottom};
    }

    private static int getPaddingByKey(ReadableMap map, String key) {
        return map.hasKey(key) ? map.getInt(key) : 0;
    }
}

File: android/src/main/java/com/rnmapbox/rnmbx/components/camera/CameraStop.kt
package com.rnmapbox.rnmbx.components.camera

import android.animation.Animator
import android.content.Context
import com.rnmapbox.rnmbx.utils.GeoJSONUtils.toPointGeometry
import com.rnmapbox.rnmbx.utils.GeoJSONUtils.toLatLng
import com.rnmapbox.rnmbx.utils.GeoJSONUtils.toLatLngBounds
import com.rnmapbox.rnmbx.utils.LatLngBounds
import com.rnmapbox.rnmbx.components.mapview.RNMBXMapView
import com.mapbox.maps.CameraOptions
import com.facebook.react.bridge.ReadableMap
import com.mapbox.geojson.FeatureCollection
import com.mapbox.maps.EdgeInsets
import com.rnmapbox.rnmbx.components.camera.constants.CameraMode
import com.rnmapbox.rnmbx.utils.LatLng

class CameraStop {
    private var mBearing: Double? = null
    private var mTilt: Double? = null
    private var mZoom: Double? = null

    private var mLatLng: LatLng? = null
    private var mBounds: LatLngBounds? = null

    private var mPaddingLeft: Int? = null
    private var mPaddingRight: Int? = null
    private var mPaddingBottom: Int? = null
    private var mPaddingTop: Int? = null

    private var mMode = CameraMode.EASE
    private var mDuration = 2000
    private var mCallback: Animator.AnimatorListener? = null

    fun setBearing(bearing: Double) {
        mBearing = bearing
    }

    fun setTilt(tilt: Double) {
        mTilt = tilt
    }

    fun setZoom(zoom: Double) {
        mZoom = zoom
    }

    fun setLatLng(latLng: LatLng?) {
        mLatLng = latLng
    }

    fun setDuration(duration: Int) {
        mDuration = duration
    }

    fun setCallback(callback: Animator.AnimatorListener?) {
        mCallback = callback
    }

    fun setPadding(
        paddingLeft: Int?,
        paddingRight: Int?,
        paddingTop: Int?,
        paddingBottom: Int?
    ) {
        mPaddingLeft = paddingLeft
        mPaddingRight = paddingRight
        mPaddingTop = paddingTop
        mPaddingBottom = paddingBottom
    }

    fun setBounds(
        bounds: LatLngBounds?,
        paddingLeft: Int?,
        paddingRight: Int?,
        paddingTop: Int?,
        paddingBottom: Int?
    ) {
        mBounds = bounds
        this.setPadding(paddingLeft, paddingRight, paddingTop, paddingBottom)
    }

    fun setMode(@CameraMode.Mode mode: Int) {
        mMode = mode
    }

    private fun convert(value: IntArray): EdgeInsets {
        val left = value[0].toDouble()
        val top = value[1].toDouble()
        val right = value[2].toDouble()
        val bottom = value[3].toDouble()
        return EdgeInsets(
            top, left, bottom, right
        )
    }

    fun toCameraUpdate(mapView: RNMBXMapView): CameraUpdateItem {
        val map = mapView.getMapboxMap()
        val currentCamera = map.cameraState

        val builder = CameraOptions.Builder()
        builder.center(currentCamera.center)
        builder.bearing(currentCamera.bearing)
        builder.padding(currentCamera.padding)
        builder.zoom(currentCamera.zoom)
        if (mBearing != null) {
            builder.bearing(mBearing)
        }
        if (mTilt != null) {
            builder.pitch(mTilt)
        }
        if (mZoom != null) {
            builder.zoom(mZoom)
        }

        val currentPadding = currentCamera.padding
        val paddingLeft: Int = mPaddingLeft ?: currentPadding.left.toInt()
        val paddingTop: Int = mPaddingTop ?: currentPadding.top.toInt()
        val paddingRight: Int = mPaddingRight ?: currentPadding.right.toInt()
        val paddingBottom: Int = mPaddingBottom ?: currentPadding.bottom.toInt()
        val cameraPadding = intArrayOf(paddingLeft, paddingTop, paddingRight, paddingBottom)
        val cameraPaddingClipped = clippedPadding(cameraPadding, mapView)
        val cameraPaddingEdgeInsets = convert(cameraPaddingClipped)
        builder.padding(cameraPaddingEdgeInsets)

        if (mLatLng != null) {
            builder.center(mLatLng!!.point)
        } else if (mBounds != null) {
            val tilt = if (mTilt != null) mTilt!! else currentCamera.pitch
            val bearing = if (mBearing != null) mBearing!! else currentCamera.bearing

            val boundsCamera = map.cameraForCoordinateBounds(
                mBounds!!.toBounds(),
                cameraPaddingEdgeInsets,
                bearing,
                tilt
            )
            builder.center(boundsCamera.center)
            builder.anchor(boundsCamera.anchor)
            builder.zoom(boundsCamera.zoom)
            builder.bearing(boundsCamera.bearing)
            builder.pitch(boundsCamera.pitch)
            builder.padding(boundsCamera.padding)
        }

        return CameraUpdateItem(map, builder.build(), mDuration, mCallback, mMode)
    }

    companion object {
        @JvmStatic
        fun fromReadableMap(
            context: Context,
            readableMap: ReadableMap,
            callback: Animator.AnimatorListener?
        ): CameraStop {
            val stop = CameraStop()

            if (readableMap.hasKey("pitch")) {
                stop.setTilt(readableMap.getDouble("pitch"))
            }

            if (readableMap.hasKey("heading")) {
                stop.setBearing(readableMap.getDouble("heading"))
            }

            if (readableMap.hasKey("centerCoordinate")) {
                val target = toPointGeometry(readableMap.getString("centerCoordinate"))
                stop.setLatLng(toLatLng(target!!))
            }

            if (readableMap.hasKey("zoom")) {
                stop.setZoom(readableMap.getDouble("zoom"))
            }

            if (readableMap.hasKey("duration")) {
                stop.setDuration(readableMap.getInt("duration"))
            }

            val metrics = context.resources.displayMetrics
            val paddingTop = getBoundsPaddingByKey(readableMap, metrics.density, "paddingTop")
            val paddingRight = getBoundsPaddingByKey(readableMap, metrics.density, "paddingRight")
            val paddingBottom = getBoundsPaddingByKey(readableMap, metrics.density, "paddingBottom")
            val paddingLeft = getBoundsPaddingByKey(readableMap, metrics.density, "paddingLeft")

            if (readableMap.hasKey("bounds")) {
                val collection = FeatureCollection.fromJson(readableMap.getString("bounds")!!)
                stop.setBounds(
                    toLatLngBounds(collection), paddingLeft, paddingRight,
                    paddingTop, paddingBottom
                )
            } else {
                stop.setPadding(paddingLeft, paddingRight, paddingTop, paddingBottom)
            }

            if (readableMap.hasKey("mode")) {
                when (readableMap.getInt("mode")) {
                    CameraMode.FLIGHT -> stop.setMode(CameraMode.FLIGHT)
                    CameraMode.LINEAR -> stop.setMode(CameraMode.LINEAR)
                    CameraMode.MOVE -> stop.setMode(CameraMode.MOVE)
                    CameraMode.NONE -> stop.setMode(CameraMode.NONE)
                    else -> stop.setMode(CameraMode.EASE)
                }
            }

            stop.setCallback(callback)

            return stop
        }

        private fun clippedPadding(padding: IntArray, mapView: RNMBXMapView): IntArray {
            val mapHeight = mapView.height
            val mapWidth = mapView.width
            val left = padding[0]
            val top = padding[1]
            val right = padding[2]
            val bottom = padding[3]
            var resultLeft = left
            var resultTop = top
            var resultRight = right
            var resultBottom = bottom
            if (top + bottom >= mapHeight) {
                val totalPadding = (top + bottom).toDouble()
                val extra =
                    totalPadding - mapHeight + 1.0 // add 1 to compensate for floating point math
                resultTop -= (top * extra / totalPadding).toInt()
                resultBottom -= (bottom * extra / totalPadding).toInt()
            }
            if (left + right >= mapWidth) {
                val totalPadding = (left + right).toDouble()
                val extra =
                    totalPadding - mapWidth + 1.0 // add 1 to compensate for floating point math
                resultLeft -= (left * extra / totalPadding).toInt()
                resultRight -= (right * extra / totalPadding).toInt()
            }
            return intArrayOf(resultLeft, resultTop, resultRight, resultBottom)
        }

        private fun getBoundsPaddingByKey(map: ReadableMap, density: Float, key: String): Int? {
            return if (map.hasKey(key)) {
                (map.getInt(key) * density).toInt()
            } else {
                null
            }
        }
    }
}